ocaml/xenopsd/lib/xenops_server.ml

```diff
@@ -2223,7 +2223,13 @@ let rec perform_atomic ~progress_callback ?subtask:_ ?result (op : atomic)
       let vifs : Vif.t list = VIF_DB.vifs id in
       let vgpus : Vgpu.t list = VGPU_DB.vgpus id in
       let vusbs : Vusb.t list = VUSB_DB.vusbs id in
-      B.VM.create_device_model t (VM_DB.read_exn id) vbds vifs vgpus vusbs
+      let vtpm : Vm.tpm option =
+        (* The uuid changes when snapshots are restored *)
+        let id = try Some (VM_DB.read_exn id) with _ -> None in
+        Option.bind id @@ fun id ->
+        match id.Vm.ty with HVM {tpm; _} -> tpm | _ -> None
+      in
+      B.VM.create_device_model t (VM_DB.read_exn id) vbds vifs vgpus vusbs vtpm
```

Comment:
Why can't the backend (xenops_server_xen.ml) itself take the vtpm out of the passed in `vm` parameter? It has a helper function for that already (`vtpm_of ~vm`), which is used in many similar functions. It can be used where currently `hvm_info.tpm` is used in `create_device_model_config`.

That way, we don't need an interface change that essentially adds an parameter with redundant data.

Reply:
`vtpm_of ~vm` is limited to what the backend can access. The vm value it has access to ultimately comes from the internal DB module: https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/xenops_server_xen.ml#L2376

This internal module has only access to the xenopsd/classic/extra directory (https://github.com/xapi-project/xen-api/blob/master/ocaml/xenopsd/xc/xenops_server_xen.ml#L188), and not the the xenopsd/classic/VM one.

I find it less invasive to add another device to the call that has already several devices passed as parameters, rather than enabling a new module to give the backend access to the information as the latter seems a breach of the self-imposed responsibilities while the former doesn't

Reply:
I'm not sure what exactly you mean, because the second argument passed to this function here is the vm record from `VM_DB.read_exn id`, which is the same thing you unpack to construct `vtpm` above.

Reply:
I see what you mean, I missed that parameter in all the parameter soup